trin-core/src/portalnet/utp.rs

```diff
@@ -1013,4 +1013,123 @@ mod tests {
         assert_eq!(reply.get_extensions()[0].bitmask[2], 0b0101_1010);
         assert_eq!(reply.get_extensions()[0].bitmask[3], 0b0000_0001);
     }
+
+    // https://github.com/ethereum/portal-network-specs/pull/127
+    // konrad uTP test vectors
+
+    #[test]
+    fn test_syn_packet() {
+        let mut response = PacketHeader::new();
+        response.set_type(Type::StSyn);
+        response.set_connection_id(10049);
+        response.set_timestamp(3384187322);
+        response.set_timestamp_difference(0);
+        response.set_wnd_size(1048576);
+        response.set_seq_nr(11884);
+        response.set_ack_nr(0);
+
+        let packet = Packet::new(response);
+        assert_eq!(
+            hex::encode(packet.0.clone()),
+            "41002741c9b699ba00000000001000002e6c0000"
+        );
+    }
+
+    #[test]
+    fn test_act_packet_no_extension() {
+        let mut response = PacketHeader::new();
+        response.set_type(Type::StState);
+        response.set_connection_id(10049);
+        response.set_timestamp(6195294);
+        response.set_timestamp_difference(916973699);
+        response.set_wnd_size(1048576);
+        response.set_seq_nr(16807);
+        response.set_ack_nr(11885);
+
+        let packet = Packet::new(response);
+        assert_eq!(
+            hex::encode(packet.0.clone()),
+            "21002741005e885e36a7e8830010000041a72e6d"
+        );
+    }
+
+    #[test]
+    fn test_act_packet_with_selective_ack_extension() {
+        let mut response = PacketHeader::new();
+        response.set_type(Type::StState);
+        response.set_connection_id(10049);
+        response.set_timestamp(6195294);
+        response.set_timestamp_difference(916973699);
+        response.set_wnd_size(1048576);
+        response.set_seq_nr(16807);
+        response.set_ack_nr(11885);
+
+        let mut incoming_buffer: BTreeMap<u16, Packet> = Default::default();
+        incoming_buffer.insert(12050, Packet::new(PacketHeader::new()));
+        incoming_buffer.insert(12081, Packet::new(PacketHeader::new()));
+
+        let mut reply = Packet::new(response);
+
+        if reply.seq_nr().wrapping_sub(12048) > 1 {
```

Comment:
I'm not closely familiar with the utp spec, but it feels a bit weird that we'd have this conditional in the test as it appears that the test should be deterministic - what's the need for the `if` statement?

Reply:
I refactored set_selective_ack and removed that cause you are correct this isn't needed here, but it is needed in the actual code